[RFC] Add `vim_call_dict_function` to API

saep opened the pull request:
This is an extension of #2979 that provides an API function for calling dict functions.
Note that this branch is based on those changes since I could reuse that code completely for the implementation of the current state.

@ZyX-I proposed this:

> Or, maybe a better idea: strip off `self` handling from `vim_call_function` completely (restore the original functionality), but additionally add `vim_call_dict_function(Object self, Boolean internal, String functionName, Array args, Error *err)` which will do the following:
> 1. Convert `self` to a dictionary: it should either be a Dictionary or a String object, in first case it is simply converted dictionary, in the last case it is expression passed to `eval()` which must evaluate to a dictionary.
> 2. If `internal` is true then check whether `functionName` is a key in `self` that refers to a function reference. If yes, replace function name with value of this reference. If no, raise an error. If `internal` is false then `functionName` is passed to `call_func` untouched.
> 3. Call `call_func` with arguments and self dictionary.
> 
> Intended usage: same as `vim_call_func`, but for constructs like `g:SyntasticLoclist.current().filter(…)`: `g:SyntasticLoclist.current()` goes into `self` as a string, arguments of `.filter()` may be different and go into `args`, rationale is the same: first string is a simple static string that does not change, arguments in this case do not require double serializing (from dictionary expected by `.filter()` to string then string to messagepack).
> 
> Additional rationale (behind the `internal` argument): this make it possible to call anonymous functions without knowing their internal name (i.e. number), thus anonymous functions can be truly anonymous like in extended-funcref branch (they are not stored in a global functions dictionary under any name there). (Or in VimL-to-lua translator, if I remember correctly.)

Here is a list of things that still have to be discussed, tested or implemented (inclusive or):
- [x] Calling with a string to be evaluated to a dict and internal set to `false`
- [x] Calling with a string to be evaluated to a dict and internal set to `true`
- [x] Calling with a msgpack-rpc dict and internal set to `false`
- [x] Calling with a msgpack-rpc dict and internal set to `true`

saep committed:
Slightly refactor vim_call_function internals

This change is made to reuse code for vim_call_dict_function.

saep committed:
Implement `vim_call_dict_function` API function

ZyX-I commented:
msgpack-rpc dict + internal=true is not possible, isn’t it? AFAIR function references cannot be passed through the API.

bfredl commented:
Please rebase on master, now #2979 is merged.

saep committed:
Use switch statement for acting upon different self types

saep committed:
Replace code snippets that unconditionally free the tmp_dict with goto statement

Some code paths were able to free the dict, even if it came from the self
argument. This would cause segfault if the self object is freed.

saep committed:
Add a few more tests

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -214,6 +226,102 @@ Object vim_call_function(String fname, Array args, Error *err)
+/// @param args Functions arguments packed in an Array
+/// @param[out] err Details of an error that may have occurred
+/// @return Result of the function call
+Object vim_call_dict_function(Object self, Boolean internal, String fname,
```
`bool`

justinmk commented:
@saep Can you rebase this?

saep commented:
So, I tried to rebase and the result can be found here: https://github.com/saep/neovim/commit/3a05715d46003c5b444a26019184fbc56a232a14

Mysteriously, this PR isn't updated with the changes I made because github doesn't know where this PR originated from. I can make another PR and reference it here, but the state isn't ready yet.

First I tried to build neovim via the nix installation instructions which do not seem to work at all (`cmakeConfigurePhase` is not available this way)
```
$ cd path/to/neovim
$ nix-shell '<nixpkgs>' -A neovim
$ cmakeConfigurePhase
build $ make
```

Then I tried adjusting the nix expression from nixpkgs, but I wasn't able to run any tests. After googling for half an hour, I gave up and created a debian VM to build and test my changes. This seemed to work initially, but when I tried to debug the failing test cases via `GDB=1 TEST_FILE=test/functional/api/vim_spec.lua TEST_FILTER=call_dict_function make functionaltest`, I managed to get these errors:

```
[  ERROR   ] test/functional/api/vim_spec.lua @ 21: api before_each
...im/.deps/usr/share/lua/5.1/nvim/child_process_stream.lua:28: ENOENT: no such file or directory

stack traceback:                 
        ...im/.deps/usr/share/lua/5.1/nvim/child_process_stream.lua:28: in function 'spawn'                                                                                       test/functional/helpers.lua:285: in function 'spawn'                                                                                                                      test/functional/helpers.lua:355: in function <test/functional/helpers.lua:319> 
```

I'm not motivated to continue anymore. Feel free to use my changes in any way you like and finish the PR.

justinmk commented:
@saep Thanks for trying. I merged it in https://github.com/neovim/neovim/pull/8353 , it seems to work.

@teto Any tips on building Nvim with nix? Is there any pending work that might improve the situation?
